From: Mike Kasick <mike@kasick.org>
Date: Tue, 18 Oct 2011 21:34:43 -0400

Relocate pmstats to avoid conflict with kexec hardboot page.

GB reserves the last page (4 kB) of physical RAM, after the ram_console,
for pm_debug_scratchpad (pmstats).  This conflicts with that memory's usage
as the hardboot page, already in use in kexec-enabled EC05 kernels.

Since pmstats uses very little of the whole 4 kB, move the scratchpad
forward to leave 1 kB of memory available for kexec.  Also requires fixing
up pmstats alloc and copy routines to not attempt to read the whole page.

diff --git a/kernel/arch/arm/mach-s5pv210/mach-victory.c b/kernel/arch/arm/mach-s5pv210/mach-victory.c
--- a/kernel/arch/arm/mach-s5pv210/mach-victory.c
+++ b/kernel/arch/arm/mach-s5pv210/mach-victory.c
@@ -3298,7 +3298,8 @@
        	ram_console_start = mi->bank[1].start + mi->bank[1].size;
 	ram_console_size = SZ_1M - SZ_4K;
 #endif
-	pm_debug_scratchpad = ram_console_start + ram_console_size;
+	/* Leave 1K at 0x57fff000 for kexec hardboot page. */
+	pm_debug_scratchpad = ram_console_start + ram_console_size + SZ_1K;
 }
 
 /* this function are used to detect s5pc110 chip version temporally */
diff --git a/kernel/arch/arm/plat-samsung/pm.c b/kernel/arch/arm/plat-samsung/pm.c
--- a/kernel/arch/arm/plat-samsung/pm.c
+++ b/kernel/arch/arm/plat-samsung/pm.c
@@ -63,8 +63,8 @@
 {
 	if (*offset != 0)
 		return 0;
-	if (len > 4096)
-		len = 4096;
+	if (len > sizeof(struct pmstats))
+		len = sizeof(struct pmstats);
 
 	if (copy_to_user(buf, file->private_data, len))
 		return -EFAULT;
@@ -89,17 +89,17 @@
 {
 	pr_info("pmstats at %08x\n", pm_debug_scratchpad);
 	if (pm_debug_scratchpad)
-		pmstats = ioremap(pm_debug_scratchpad, 4096);
+		pmstats = ioremap(pm_debug_scratchpad, sizeof(struct pmstats));
 	else
-		pmstats = kzalloc(4096, GFP_ATOMIC);
+		pmstats = kzalloc(sizeof(struct pmstats), GFP_ATOMIC);
 
 	if (!memcmp(pmstats->magic, PMSTATS_MAGIC, 16)) {
-		pmstats_last = kzalloc(4096, GFP_ATOMIC);
+		pmstats_last = kzalloc(sizeof(struct pmstats), GFP_ATOMIC);
 		if (pmstats_last)
-			memcpy(pmstats_last, pmstats, 4096);
+			memcpy(pmstats_last, pmstats, sizeof(struct pmstats));
 	}
 
-	memset(pmstats, 0, 4096);
+	memset(pmstats, 0, sizeof(struct pmstats));
 	memcpy(pmstats->magic, PMSTATS_MAGIC, 16);
 
 	debugfs_create_file("pmstats", 0444, NULL, pmstats, &pmstats_ops);
